### PR TITLE
[SYSTEMML-1493] [SYSTEMML-1500] Added TanH and Euclidean Loss in Caffe2DML

### PR DESCRIPTION
**Pull request by niketanpansare**
- Added [the reference documentation](http://niketanpansare.github.io/systemml/reference-guide-caffe2dml). 
- Added a new layer softmax_loss.dml
- Added compute_loss_accuracy to l2_loss.dml
- Updated tanh layer to invoke newly added builtin function

@dusenberrymw can you please review this PR ?

@bertholdreinwald  @prithvirajsen

### CODE REVIEW

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2124/<h2>Failed Tests: <span class='status-failure'>8</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2124/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>8</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2124/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.codegen/CellwiseTmplTest/testCodegenCellwiseRewrite13_sp/'><strong>org.apache.sysml.test.integration.functions.codegen.CellwiseTmplTest.testCodegenCellwiseRewrite13_sp</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2124/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.codegen/CellwiseTmplTest/testCodegenCellwise6/'><strong>org.apache.sysml.test.integration.functions.codegen.CellwiseTmplTest.testCodegenCellwise6</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2124/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.codegen/CompressedRowAggregateTest/testCompressedRowAggregateMainSparseRandCP/'><strong>org.apache.sysml.test.integration.functions.codegen.CompressedRowAggregateTest.testCompressedRowAggregateMainSparseRandCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2124/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.scripts.nn/NNTest/testNNLibrary/'><strong>org.apache.sysml.test.integration.scripts.nn.NNTest.testNNLibrary</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2124/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.gdfo/GDFOLinregDS/testGDFOLinregDSSparseMR/'><strong>org.apache.sysml.test.integration.functions.gdfo.GDFOLinregDS.testGDFOLinregDSSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2124/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.gdfo/GDFOLinregDS/testGDFOLinregDSDenseMR/'><strong>org.apache.sysml.test.integration.functions.gdfo.GDFOLinregDS.testGDFOLinregDSDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2124/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.gdfo/GDFOLinregDSsimpl/testGDFOLinregDSSparseMR/'><strong>org.apache.sysml.test.integration.functions.gdfo.GDFOLinregDSsimpl.testGDFOLinregDSSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2124/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.gdfo/GDFOLinregDSsimpl/testGDFOLinregDSDenseMR/'><strong>org.apache.sysml.test.integration.functions.gdfo.GDFOLinregDSsimpl.testGDFOLinregDSDenseMR</strong></a></li></ul>

File: scripts/nn/layers/l2_loss.dml [hunk 1]
```diff
@@ -70,3 +70,21 @@ backward = function(matrix[double] pred, matrix[double] y)
   dpred = (pred-y) / N
 }
 
+compute_loss_accuracy = function(matrix[double] probs, matrix[double] y)
```
Comment (dusenberrymw) [inline scripts/nn/layers/l2_loss.dml]: what is this for?  if it's is supposed to return the loss and the accuracy, let's just have Caffe2DML call the appropriate loss function and separately an accuracy function.  I don't think it's a good idea to pollute the `nn` library to account for Caffe issues.

File: scripts/nn/layers/softmax2d_loss.dml [hunk 2]
```diff
@@ -0,0 +1,80 @@
+#-------------------------------------------------------------
```
Comment (dusenberrymw) [inline scripts/nn/layers/softmax2d_loss.dml]: we already have the softmax2d and the associated cross_entropy_loss2d layers in the `nn` library.  let's have Caffe2DML simply use those two layers, rather than adding this wrapper layer.

File: scripts/nn/layers/softmax_loss.dml [hunk 3]
```diff
@@ -0,0 +1,108 @@
+#-------------------------------------------------------------
```
Comment (dusenberrymw) [inline scripts/nn/layers/softmax_loss.dml]: we already have softmax and cross_entropy_loss layers.  let's have Caffe2DML simply use those two layers, rather than addin this new wrapper.

File: scripts/nn/layers/tanh.dml [hunk 4]
```diff
@@ -29,21 +29,13 @@ forward = function(matrix[double] X)
   /*
    * Computes the forward pass for a tanh nonlinearity layer.
    *
-   *   ```
-   *   tanh(x) = (e^x - e^-x) / (e^x + e^-x)
-   *           = 2 * sigmoid(2x) - 1
-   *   ```
-   *
```
Comment (dusenberrymw) [inline scripts/nn/layers/tanh.dml]: I would leave this comment in, as it gives the math for a tanh function.

File: scripts/nn/layers/tanh.dml [hunk 5]
```diff
@@ -29,21 +29,13 @@ forward = function(matrix[double] X)
...
    *  - X: Inputs, of shape (any, any).
    *
    * Outputs:
    *  - out: Outputs, of same shape as `X`.
    */
-  # out = (exp(X) - exp(-X)) / (exp(X) + exp(-X))
-  # Simplification of the above formulation to use the sigmoid function:
-  sigma2X = sigmoid::forward(2*X)
-  out = 2*sigma2X - 1
+  out = tanh(X)
```
Comment (dusenberrymw) [inline scripts/nn/layers/tanh.dml]: if we're going to swap this out, can you first make a copy of this tanh layer into the `nn/test` folder, and add a unit test comparing the old and new versions in `nn/test/test.dml`?

File: scripts/nn/layers/l2_loss.dml (hunk 1 above)
Comment (niketanpansare) [reply scripts/nn/layers/l2_loss.dml]: The reason I added a separate loss and accuracy function is to allow DML writers flexibility to modify them for their problem. This goes in direction of making Caffe2DML a thin wrapper :)

File: scripts/nn/layers/tanh.dml (hunk 5 above)
Comment (niketanpansare) [reply scripts/nn/layers/tanh.dml]: Done.

File: scripts/nn/layers/tanh.dml (hunk 4 above)
Comment (niketanpansare) [reply scripts/nn/layers/tanh.dml]: Done.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2128/

**Comment by niketanpansare** [comment]
@dusenberrymw any additional comments ?

File: docs/reference-guide-caffe2dml.md [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+  name: "tanh1"
+  type: "TanH"
+  bottom: "conv1"
+  top: "conv1"
+}
+```
+
+### Sigmoid Layer
+
+Invokes [nn/layers/tanh.dml](https://github.com/apache/systemml/blob/master/scripts/nn/layers/sigmoid.dml) layer.
```
Comment (dusenberrymw) [inline docs/reference-guide-caffe2dml.md]: minor: `nn/layers/tanh.dml` -> `nn/layers/sigmoid.dml`

File: docs/reference-guide-caffe2dml.md [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+  name: "concat_d5cc_u5a-b"
+  type: "Concat"
+  bottom: "u5a"
+  bottom: "d5c"
+  top: "u5b"
+}
+```
+
+### Softmax Layer
+
```
Comment (dusenberrymw) [inline docs/reference-guide-caffe2dml.md]: minor: add the `nn/layers/softmax.dml` file link here to be consistent

File: docs/reference-guide-caffe2dml.md [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+Loss drives learning by comparing an output to a target and assigning cost to minimize. 
+The loss itself is computed by the forward pass and the gradient w.r.t. to the loss is computed by the backward pass.
+
+### Softmax with Loss Layer
+
+The softmax loss layer computes the multinomial logistic loss of the softmax of its inputs. 
+It’s conceptually identical to a softmax layer followed by a multinomial logistic loss layer, but provides a more numerically stable gradient.
+
+Invokes [nn/layers/softmax_loss.dml](https://github.com/apache/systemml/blob/master/scripts/nn/layers/softmax_loss.dml) or
+[nn/layers/softmax2d_loss.dml](https://github.com/apache/systemml/blob/master/scripts/nn/layers/softmax2d_loss.dml) layer.
```
Comment (dusenberrymw) [inline docs/reference-guide-caffe2dml.md]: minor: update these links to use a combination of `nn/layers/softmax.dml` and `nn/layers/cross_entropy_loss.dml`, or the 2D equivalents.

**Comment by niketanpansare** [comment]
Thanks @dusenberrymw  ... I will merge :+1:

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2139/
